fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/codec/LazyCloseableTest.java

```diff
@@ -0,0 +1,243 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class LazyCloseableTest {
+
+    @Test
+    void testOpensLazilyExactlyOnce() throws IOException {
+        final AtomicInteger openCounter = new AtomicInteger(0);
+        final AtomicInteger closeCounter = new AtomicInteger(0);
+        try (LazyCloseable<CountingCloseable> opener = LazyCloseable.supply(
+                () -> new CountingCloseable(openCounter, closeCounter))) {
+            assertEquals(0, openCounter.get());
+            assertEquals(1, opener.get().openCounts);
+            assertEquals(1, opener.get().openCounts);
+            assertEquals(1, opener.getUnchecked().openCounts);
+            assertSame(opener.get(), opener.get());
+            assertEquals(1, openCounter.get());
+        }
+        assertEquals(1, closeCounter.get());
+    }
+
+    @Test
+    void testOpensLazilyExactlyOnceThreaded() {
+        AtomicInteger starts = new AtomicInteger(0);
+        AtomicInteger ends = new AtomicInteger(0);
+        AtomicInteger opens = new AtomicInteger(0);
+        AtomicInteger closes = new AtomicInteger(0);
+        final int concurrency = 100;
+        final LazyCloseable<CountingCloseable> opener = LazyCloseable.supply(() -> {
+            starts.incrementAndGet();
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e) {
+                throw new AssertionError("Timed out waiting for latch");
+            }
+            ends.incrementAndGet();
+            return new CountingCloseable(opens, closes);
+        });
+        ConcurrentHashMap<Thread, Integer> threads = new ConcurrentHashMap<>();
+        final List<CountingCloseable> allValues = IntStream.range(0, concurrency)
+                .parallel()
+                .mapToObj(i -> {
+                    threads.compute(Thread.currentThread(), (k, v) -> v == null ? 1 : v + 1);
+                    return opener.getUnchecked();
+                })
+                .collect(Collectors.toList());
+        MatcherAssert.assertThat(allValues, Matchers.everyItem(Matchers.sameInstance(allValues.get(0))));
+        // Stream.parallel doesn't guarantee that it will run in 100 parallel threads
+        MatcherAssert.assertThat(threads.keySet(), Matchers.hasSize(Matchers.greaterThan(10)));
+    }
+
+    @Test
+    void testForkJoinPoolDeadlock() throws ExecutionException, InterruptedException, TimeoutException {
```

Comment:
I was inspired to add this test, and the version on `LazyOpenerTest`, since this seems particularly important for our use case.